### PR TITLE
refactor(core): warn about duplicated keys when using built-in @for

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This commit a new check that warn users about duplicated keys detected given a tracking expression and a collection iterated over with @for. Duplicated keys should be avoided as those are more expensive to manage and can result in incorrect UI display.

### CODE REVIEW

File: packages/core/src/errors.ts [hunk 1]
```diff
@@ -127,6 +127,9 @@ export const enum RuntimeErrorCode {
   // Runtime dependency tracker errors
   RUNTIME_DEPS_INVALID_IMPORTED_TYPE = 1000,
   RUNTIME_DEPS_ORPHAN_COMPONENT = 1001,
+
+  // Repeater errors (1050-1099 range)
+  LOOP_TRACK_DUPLICATE_KEYS = 1050,
```
Comment (JeanMeche) [inline packages/core/src/errors.ts:132]: We have a conflict with the forms package with code > 1000.

Comment (pkozlowski-opensource) [reply packages/core/src/errors.ts]: Oh, good catch, thnx!

File: packages/core/src/render3/instructions/control_flow.ts [hunk 2]
```diff
@@ -253,6 +254,35 @@ class LiveCollectionLContainerImpl extends
...
+
+  if (duplicatedKeysMsg.length > 0) {
+    const message = formatRuntimeError(
+        RuntimeErrorCode.LOOP_TRACK_DUPLICATE_KEYS,
+        'The provided track expression resulted in duplicated keys for a given collection. ' +
+            'Adjust the tracking expression such that it uniquely identifies all the items in the collection. ' +
+            'Duplicated keys were: \n' + duplicatedKeysMsg.join(', \n') + '.');
+
+    // tslint:disable-next-line:no-console
+    console.warn(message);
```
Comment (JeanMeche) [inline packages/core/src/render3/instructions/control_flow.ts:282]: It there a way we could rely on the `Console` token defined at `packages/core/src/console.ts` ?

Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/control_flow.ts:282]: We've got access to LView so could inject it from there. I'm actually not clear about the pattern we should follow as we are randomly talking directly to `console` or injecting `Console`.

**Review by crisbeto — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Comment by atscott** [comment]
This PR was merged into the repository by commit e3696ad0d6007fe48dcf2b7ee0ad52bac767de1b.

**Comment by atscott** [comment]
Added cleanup to investigate failure in cl/623563717

**Comment by pkozlowski-opensource** [comment]
Thnx @atscott - this change can indeed cause failures in case of using an iterator (that can be iterated over only once). Going to add a test for this case and move the logic to the list reconciliation itself so we are not iterating twice.

**Comment by pkozlowski-opensource** [comment]
Looking more into G3 failures, there is one where a test fails because of a new warning. Need to look into it more.

**Comment by pkozlowski-opensource** [comment]
G3 cleanup is done, marking for merge.

**Comment by pkozlowski-opensource** [comment]
This PR was merged into the repository by commit 9f7b9ed87a671d4df5d9e519d57a78d718f44524.

**Comment by alxhub** [comment]
This PR was merged into the repository by commit e67a73068802dd33d768c0910e74dd1de8341b2a.

**Pull request closed without merging**